### PR TITLE
feat: add goose adapter (SQLite sessions.db + legacy JSONL)

### PR DESCRIPTION
**Pull request by jelloeater-agent**
## Summary

Adds the `goose` adapter for Block's Goose CLI (github.com/aaif-goose/goose): lossless ingest of sessions from both storage generations into pond's canonical model.

- **SQLite generation** (goose >= 1.10): `<data-dir>/sessions/sessions.db` — `sessions`/`messages` tables ingested with the upstream `ORDER BY (normalized created_timestamp), id` ordering.
- **Legacy JSONL generation** (pre-1.10): `<data-dir>/sessions/<id>.jsonl` — line-1 metadata + message lines; read so pre-migration files never strand (`session-movement-complete`).
- Sessions present in both sources resolve to the DB; the legacy copy is counted `Superseded` (never silently dropped).
- Content-block inventory mapped: text/image/thinking to conversational parts, toolRequest/toolResponse/tool confirmation/actionRequired/systemNotification/error to injected parts (provenance override via `metadata.userVisible`/`turnContext`), unknown future block types to lossless compact-repr carriers.
- Source-agent taxonomy from `session_type`: `goose`, `goose/subagent`, `goose/scheduled`, `goose/hidden`, `goose/terminal`, `goose/gateway`, `goose/acp`. Lineage via `parent_session_id`.
- Ingest-only: `restore_unsupported` — goose reads its own sessions directly; a restore face via `goose session import` is a follow-up.

Full format archaeology and the 11-row decision table: `docs/adapters/goose.md`.

## Fixture note

The fixture DB (`tests/fixtures/adapter/goose/data/sessions/`) was **generated**, not live-captured — goose needs LLM provider credentials to produce sessions, and none were available in the capture environment. Schema is verbatim from upstream `session_manager.rs`; rows exercise multi-turn text, tool request+response, sub_agent lineage, thinking, and scheduled session types. Legacy JSONL samples are documented synthetic-supplementary (the writer no longer emits them). A live sandbox self-capture is the follow-up that would upgrade the fixture to captured ground truth.

## Verification

- `cargo fmt --check` clean
- `cargo clippy --all-targets -- -D warnings` clean
- `cargo test`: lib 331/331, integration 110/110 (2 pre-existing ignored), doc 0

## PR source note

Opened as a draft from the `jelloeater-agent` fork: this machine's only GitHub identity has no push access to `tenequm/pond`. When an authorized identity is available, re-run the same-repo flow (`git push origin feat/goose-adapter` then `gh pr create --draft`) or a maintainer can push this branch directly to enable the same-repo CI leg.

### CODE REVIEW

File: packages/pond/src/adapter/goose.rs [hunk 1]
```diff
@@ -0,0 +1,1790 @@
...
+            if has_sessions_dir(&data) {
+                return Some(data);
+            }
+        }
+    }
+    // XDG_DATA_HOME / fallback
+    let xdg = std::env::var_os("XDG_DATA_HOME")
+        .map(PathBuf::from)
+        .unwrap_or_else(|| env.home.join(".local").join("share"));
+    let data = xdg.join("goose").join("data");
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:167]: **Required.** The data dir is wrong on all three platforms, so `probe_default` can never detect a real install. Upstream `paths.rs` uses `etcetera::choose_app_strategy`, which is the Xdg strategy on macOS as well as Linux, so the real dir is `~/.local/share/goose` on both - no `Library/Application Support` arm, and no `/data` suffix: goose appends `data` only under `GOOSE_PATH_ROOT`. Windows is `%APPDATA%\Block\goose\data`, and there is no Windows arm here although the spec doc claims one.

Drop the `.join("data")`, delete the macOS arm below, add the Windows arm, and fix the README row to match. Playbook Phase B step 7, last paragraph: `pond adapters list` on a real install must show `detected` - that run is what would have caught this.

File: packages/pond/src/adapter/goose.rs [hunk 2]
```diff
@@ -0,0 +1,1790 @@
...
+        let root = PathBuf::from(root);
+        if root.is_absolute() {
+            let data = root.join("data");
+            if has_sessions_dir(&data) {
+                return Some(data);
+            }
+        }
+    }
+    // XDG_DATA_HOME / fallback
+    let xdg = std::env::var_os("XDG_DATA_HOME")
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:164]: **Required.** No `std::env` reads inside an adapter (Phase B step 3): the probe reads only the injected `Env`, because env lookups cannot be injected and their tests then fail or skip on any box that sets the variable. `probe_default_respects_data_dir` below hard-fails wherever `XDG_DATA_HOME` is set. A relocated root is configured as an explicit `path`; `GOOSE_PATH_ROOT` and `XDG_DATA_HOME` are the discovery layer's concern (#148).

File: packages/pond/tests/fixtures/README.md [hunk 3]
```diff
@@ -329,6 +331,35 @@ sample tree.
...
+- Source-of-truth layout: `<data-dir>/sessions/` under
+  `~/.local/share/goose/data/` (Linux), `~/Library/Application Support/Block/goose/data/`
+  (macOS), `%APPDATA%\Block\goose\data` (Windows).
+  - `sessions.db` - WAL, `sessions`/`messages` tables.
+    Schema: github.com/aaif-goose/goose/blob/main/crates/goose/src/session/session_manager.rs ~L991-1090.
+    Checkpointed to rollback-journal `.db` here so the fixture is a single file.
+  - `*.jsonl` - legacy pre-1.10 format, one per session. Filename stem = session id.
+    Line 1 metadata; lines 2+ messages `{role, created, content}`.
+
+- DB sample: GENERATED via the sqlite3 CLI at fixture-capture time, not
```
Comment (tenequm) [inline packages/pond/tests/fixtures/README.md:349]: **Required.** The fixture is 100% generated. Playbook Phase A step 2: fixtures come from sandboxed self-capture only; generator synthetics are supplementary edge-case rows, never the round-trip ground truth. Run goose under a throwaway `HOME` with a local or free provider (that clears the credential gate the PR body describes), and script sessions that cover the shapes the step lists - a failed tool call, an unfinished call, an empty session, an image, an error block, an unknown block type, a non-user/assistant role, millisecond timestamps. Then add the per-file row census this section needs; every count a test asserts derives from it.

File: packages/pond/src/adapter/goose.rs [hunk 4]
```diff
@@ -0,0 +1,1790 @@
...
+        if let (session_id, Some(ts)) = row {
+            map.insert(session_id, ts * 1_000_000);
+        }
+    }
+    Some(map)
+}
+
+fn peek_jsonl_watermark(path: &Path) -> Option<i64> {
+    // Read the last non-metadata line for the newest timestamp.
+    let content = std::fs::read_to_string(path).ok()?;
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:416]: **Required.** `read_to_string` of the whole file, here and in `read_legacy_session` (~line 1117): the legacy path does not use `jsonl.rs` at all - no `parse_bounded`, no `PEEK_TAIL_CAP` tail peek (`adapter-bounded-values`, Phase B step 1). This also makes decision-table row 10 ("tail-peek of the last line") untrue: it scans every line for a MAX. Route the legacy generation through `jsonl.rs`, and add the `peek_metering::measure` read-budget test that `grok_build.rs` and `letta_code.rs` carry (Phase B step 5).

File: packages/pond/src/adapter/goose.rs [hunk 5]
```diff
@@ -0,0 +1,1790 @@
...
+                    ) {
+                        parts.push(part);
+                        part_ord += 1;
+                    }
+                }
+                Message::User {
+                    id: msg_id,
+                    session_id: session_id.to_owned(),
+                    timestamp,
+                    options: ProviderOptions::new(),
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:1270]: **Required.** Legacy messages get an empty `ProviderOptions` (here, 1292, 1302) - no `options.source.raw_record`, so `id`, `created`, and any unmapped field are unrecoverable. `model-lossless-projection` fails for the whole legacy generation. Use `source_options` + `raw_record` as the DB path does.

File: packages/pond/src/adapter/goose.rs [hunk 6]
```diff
@@ -0,0 +1,1790 @@
...
+                }
+            }
+            Message::Assistant {
+                id: message_id,
+                session_id: session_id.to_owned(),
+                timestamp,
+                options,
+            }
+        }
+        _ => {
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:767]: **Required.** A non-`user`/`assistant` role goes through `extract_str_from_content`, which keeps only the first `text` block and discards the rest - and this arm emits zero `Part`s. Same in the legacy reader (~line 1295), where there is no `raw_record` either, so it is real data loss there. Placement rule 3 (spec 6.5): unknown roles land verbatim in a lossless carrier, never on the floor.

File: packages/pond/src/adapter/goose.rs [hunk 7]
```diff
@@ -0,0 +1,1790 @@
...
+/// Map a single content block to a Part, or None for skipped/unsupported types.
+fn map_content_block(
+    session_id: &str,
+    message_id: &str,
+    ordinal: usize,
+    block: &Value,
+    metadata: &Value,
+    tool_name_map: &HashMap<String, String>,
+) -> Option<Part> {
+    let block_type = block.get("type").and_then(Value::as_str)?;
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:795]: **Required.** A block with no `type` is dropped by this `?` with no counted skip and no error; the same happens to unparseable `content_json` at ~line 719 (`.ok().unwrap_or_default()`). `adapter-integrity-no-silent-drops`: a record the adapter cannot ingest is a typed error or a counted `SkipReason` naming the file and the fix.

File: docs/adapters/goose.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+- `crates/goose-provider-types/src/conversation/tool_result_serde.rs` - the `ToolResult` serde shape (`status`, `error`, `value`, per-tool result codecs).
+- `crates/goose/src/session/legacy.rs` - the pre-1.10 JSONL writer (line 1 session metadata, lines 2+ messages).
+- `crates/goose/src/config/paths.rs` - where the data directory lives (default and env overrides).
+- `crates/goose-cli/src/commands/session.rs` - CLI-facing session identity and subagent/scheduled classification.
+
+Third-party reference (NON-authoritative, used only to cross-check shape archaeology; the on-disk code in `packages/pond/src/adapter/goose.rs` is authoritative for extraction behavior):
+
+- `github.com/vshulcz/deja-vu/blob/main/internal/sources/goose.go`
+
+## Decision table
```
Comment (tenequm) [inline docs/adapters/goose.md:20]: **Required.** Every decision-table answer cites evidence - an upstream permalink or an observed line in the capture (Phase A step 3) - and the doc carries a field-history section from `git log --follow -S` on the writer files (steps 1 and 4). Neither is here; compare `docs/adapters/letta-code.md`, whose rows each cite `file:line`. Two content fixes while you are in it: row 8 should declare `usage_ledger` (per-session token/cost rows exist upstream and are not ingested), and `frontendToolRequest` was removed upstream on 2026-08-28 (PR #11645) - one day after the `Last verified` line - so label that arm as legacy.

File: docs/adapters/goose.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+| # | Decision | Resolution |
+|---|----------|------------|
+| 1 | source_agent brand | `goose`; kind-subpath taxonomy keyed on `session_type`: user/missing -> `goose`, `sub_agent` -> `goose/sub-agent`, `scheduled` -> `goose/scheduled`, `hidden` -> `goose/hidden`, `terminal` -> `goose/terminal`, plus `goose/gateway`, `goose/acp`. |
+| 2 | Session identity | DB session `id` verbatim. Message id = `<session_id>:<messages.id>`. Legacy message id = own `id` field, else `<session>:line<N>` from `options.source.message_id` parity. |
```
Comment (tenequm) [inline docs/adapters/goose.md:25]: `options.source.message_id` is never written anywhere in the adapter - legacy messages get no `options.source` at all. Either write it or drop the claim.

File: docs/adapters/goose.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+| 7 | Lineage | `parent_session_id` from the DB `sessions` row; mapped to `options.goose.relation` (`child`/`root`). |
+| 8 | Deliberate non-capture | Provider/extension inventory stored in non-session tables is not captured; only `sessions` and `messages` are ingressed. |
+| 9 | Restore face | Ingest-only: `sessions.db` / legacy JSONL are read but not restored to disk. Follow-up sessions re-imported via the `pond` session import envelope. |
+| 10 | Freshness | DB: `MAX(normalized created_timestamp)` microseconds grouped over `messages`. Legacy: tail-peek of the last line's `created` timestamp; empty file / source with no usable watermark -> `SourceWatermark::Empty`. Where both generations hold the same session id, the DB wins (legacy counted as superseded). |
+| 11 | Windows | Data dir under `%APPDATA%\Block\goose\data` probe arm; path components joined with `.join()`; JSONL lines trim `\r`; timestamps parsed as UTC. |
```
Comment (tenequm) [inline docs/adapters/goose.md:34]: There is no `%APPDATA%` probe arm in `resolve_data_dir`. Row 11 has to describe what the code does, and the code has to do what upstream does (see the discovery comment).

File: packages/pond/src/adapter/goose.rs [hunk 11]
```diff
@@ -0,0 +1,1790 @@
...
+//! - **Legacy JSONL** (pre-1.10): `<data-dir>/sessions/<id>.jsonl`.
+//!   Line 1 session metadata; lines 2+ messages.
+//!
+//! Both are read. Session ids present in the DB supersede legacy copies
+//! (counted, never silently dropped).
+//!
+//! Identity: DB session `id` verbatim; message id = `<session_id>:<rowid>`.
+//! Legacy message id = own `id` field or `<session>:line<N>`.
+//!
+//! Source agent taxonomy: `goose` (default/user), `goose/subagent`
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:17]: The code emits `goose/sub-agent` (`session_type_to_source_agent` replaces `_` with `-`), which the spec doc and the integration test agree on. This comment and the PR body say `goose/subagent`. Fix both.

File: packages/pond/src/adapter/goose.rs [hunk 12]
```diff
@@ -0,0 +1,1790 @@
...
+                )));
+                continue;
+            }
+        };
+
+        let msg_id = msg_value
+            .get("id")
+            .and_then(Value::as_str)
+            .map(ToOwned::to_owned)
+            .unwrap_or_else(|| format!("{}:line{}", session_id, line_idx + 1));
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:1230]: Off by one: the fallback id uses `line_idx + 1` while the parse error eleven lines up reports the same record as `line_idx + 2` (line 1 is metadata). File line 2 gets id `...:line1`.

File: packages/pond/src/adapter/goose.rs [hunk 13]
```diff
@@ -0,0 +1,1790 @@
...
+    };
+
+    let mut events = Vec::with_capacity(parts.len() + 1);
+    events.push(IngestEvent::Message(message));
+    events.extend(parts.into_iter().map(IngestEvent::Part));
+    Ok(events)
+}
+
+/// Map a single content block to a Part, or None for skipped/unsupported types.
+fn map_content_block(
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:787]: This function is 307 lines and repeats the six-field `Part { session_id, id, message_id, ordinal, .. }` literal fourteen times; a local `part(kind, provenance, options)` closure halves it. The `Some("user")` / `Some("assistant")` block loops are byte-identical and duplicated across `db_message_events` and `read_legacy_session` (four copies), and the `frontendToolRequest` arm (~line 1014) duplicates the `toolRequest` success path verbatim. Roughly 200-250 lines come out with no behavior change; this adapter is currently the largest SQLite adapter after opencode with the lowest test-to-code ratio.

File: packages/pond/src/adapter/goose.rs [hunk 14]
```diff
@@ -0,0 +1,1790 @@
...
+                    AdapterYield::Event(event) => out.push(event),
+                    other => panic!("unexpected non-event yield: {other:?}"),
+                }
+            }
+            out
+        })
+    }
+
+    #[test]
+    fn goose_schema_matches_upstream() {
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:1483]: Tautology: this creates a DB from the test's own `GOOSE_SCHEMA` constant and asserts the tables it just created exist - it verifies nothing about upstream. `parse_ts_epoch_seconds_normalizes_millis` (~1539) asserts only `is_some()` twice and never checks normalization, and the `tool_correlation_*` tests reimplement the lookup in the test body and assert on `HashMap::get` without exercising the `toolResponse` arm. Each decision-table row that involved a choice gets a real test (Phase B step 5).

File: packages/pond/tests/integration/adapter/goose.rs [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+// 4 DB sessions + 2 legacy JSONL sessions (disjoint ids from DB).
+const EXPECTED_SESSIONS: usize = 6;
+
+fn conformance() -> Conformance<'static> {
+    Conformance {
+        factory: &GooseFactory,
+        fixture_root: Path::new(FIXTURE_ROOT),
+        expected_sessions: EXPECTED_SESSIONS,
+        resync_rereads: &[],
```
Comment (tenequm) [inline packages/pond/tests/integration/adapter/goose.rs:20]: The suite asserts one session's brand and one parent id. `hermes.rs` (also SQLite) is the model: taxonomy census over all sessions, lineage for each kind, project fallback chain, and tool/multimodal parts surviving the store. Add those, plus tool call/result pair survival, a thinking part, and DB-vs-legacy coexistence (a session id present in both resolving to the DB with the legacy copy counted `Superseded`).

File: packages/pond/src/adapter/goose.rs [hunk 16]
```diff
@@ -0,0 +1,1790 @@
...
+        "session has no parseable created_at or updated_at timestamp",
+    )));
+    None
+}
+
+fn classify_source_agent(row: &Value) -> String {
+    let session_type = row
+        .get("session_type")
+        .and_then(Value::as_str)
+        .unwrap_or("user");
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:594]: `.unwrap_or("user")` is a synthesized default on a `NOT NULL DEFAULT 'user'` column that cannot be missing - drop it (`model-no-synthesis`). Four lines down, `session_type_to_source_agent` maps any string to `goose/<value>` unvalidated; a value containing `/` or whitespace produces a malformed `source_agent`. Bound it to the seven upstream variants and treat anything else as an unsupported record.

File: packages/pond/src/adapter/goose.rs [hunk 17]
```diff
@@ -0,0 +1,1790 @@
...
+            Ok(conn) => {
+                let ids = match list_session_ids(&conn, &db_path) {
+                    Ok(ids) => ids,
+                    Err(error) => {
+                        errors.push(error);
+                        Vec::new()
+                    }
+                };
+                let watermarks = if peek {
+                    session_watermarks(&conn).unwrap_or_default()
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:323]: `unwrap_or_default()` turns a DB error into "every session is stale" with no diagnostic. Surface it as a typed error naming the file.

File: packages/pond/src/adapter/goose.rs [hunk 18]
```diff
@@ -0,0 +1,1790 @@
...
+            NAME,
+            message_id.to_owned(),
+            "message has no parseable created_timestamp",
+        )),
+    }
+}
+
+// -- Small helpers ----------------------------------------------------------
+
+fn message_options(row: &Value, rowid: i64) -> ProviderOptions {
```
Comment (tenequm) [inline packages/pond/src/adapter/goose.rs:1371]: Hand-rolls `options.source` with a non-standard `id` field instead of calling the shared `source_options` helper (Phase B step 1).

File: packages/pond/tests/integration/adapter/mod.rs [hunk 19]
```diff
@@ -28,6 +28,7 @@ use tempfile::TempDir;
 mod claude_ai_export;
 mod claude_code;
 mod grok_build;
+mod goose;
```
Comment (tenequm) [inline packages/pond/tests/integration/adapter/mod.rs:31]: Nit: alphabetical order - `goose` goes before `grok_build`, here and in `src/adapter/mod.rs`.

File: README.md [hunk 20]
```diff
@@ -268,6 +268,7 @@ One adapter per harness, in `pond adapters` discovery order; `Reads` is the path
 | `openclaw` | openclaw, `~/.openclaw` | 2026-05-13 |
 | `nanoclaw` | nanoclaw, `~/nanoclaw` (the install root holding `data/v2-sessions`) | 2026-05-14 |
 | `hermes` | Hermes Agent, `~/.hermes` (`state.db` per profile) | 2026-07-23 |
+| `goose` | Goose (Block), `~/.local/share/goose/data` (SQLite DB + legacy JSONL) | 2026-08-25 |
```
Comment (tenequm) [inline README.md:271]: The `Reads` path follows the discovery fix: `~/.local/share/goose` (Linux and macOS), no `/data`.

**Review by tenequm — CHANGES_REQUESTED** [review]
Thanks for this. The upstream archaeology is the expensive half and it is right: the repo (block/goose -> aaif-goose/goose), all six writer files, both column lists (29 session / 9 message columns, column-for-column against `create_schema()`), the `session_type` enum, the seconds-not-millis `created_timestamp` and its normalizer, and the pre-1.10 JSONL story all check out against the real writer. Not merging yet, because the parts of the playbook that need a real install were not run, and they are the parts that decide whether this adapter does anything for a user.

What has to change, each with an inline comment pointing at the playbook step it comes from (`.agents/skills/add-adapter/SKILL.md`, also the `/add-adapter` skill):

1. Discovery is wrong on every platform - the adapter cannot find a real goose install anywhere (Phase B step 3 + the run-as-a-user check in step 7).
2. The fixture is fully generated, with none of the edge shapes the playbook lists (Phase A step 2). It needs a sandboxed self-capture; a local or free provider clears the credential gate.
3. The legacy JSONL path bypasses `jsonl.rs` (unbounded reads, no tail peek, no read-budget test) and its messages carry no `raw_record` (Phase B steps 1 and 5).
4. The spec doc has no evidence citations and no field-history section (Phase A steps 3 and 4), and disagrees with the code in four places.
5. Silent drops, a system-role data loss, an off-by-one, and three tautological tests - all inline.

Also missing from the PR itself: a `## Release note` section (CONTRIBUTING), the harness roster in `docs/site/src/pages/index.mdx`, and the `## Adapter evidence` section CONTRIBUTING now asks adapter PRs for (#200): fixture provenance plus the output of `pond sync goose --path <fixture>`, the re-sync skipping fresh, a default `pond search` hit, and `pond adapters list` showing `detected` on a real install. Paste those and most of the review answers itself.

On the draft note about push access: a fork PR needs none - this PR is already in the right place, and CI's Windows leg runs when I push the branch or on merge. If there is no movement in about a week I will finish it on this branch, keeping you as author.
